### DISCUSSION TITLE
Proper way to annotate a function that could return different models?

### DISCUSSION THREAD

**Discussion by samuel-odinware** [pydantic/pydantic#7755] [Question]
As an example, I need to parse spreadsheet data. I have multiple spreadsheets with different data structures, so I was looking to make a reusable parser. This was my initial thought for a function that allows me to parse the row data or collect errors to return to the user. This seems to break any IDE support.

```python
"""Parse row to data model."""

from typing import Any, TypeVar

from openpyxl.cell import Cell
from pydantic import BaseModel, ValidationError


RowModelT = TypeVar("RowModelT", bound=BaseModel)


def parse_row(
    row: tuple[Cell], columns_map: dict[str, dict[str, Any]], row_model: RowModelT
) -> tuple[RowModelT | None, Exception | None]:
    """Parse row data into a row model.

    Args:
    ----
        row (tuple[Cell]): Row containing `Cell` objects to be parsed.
        columns_map (dict[str, dict[str, Any]]): Map of column letter to matching field names.
        row_model (RowModelT): Model to initialize with data.

    Returns:
    -------
        tuple[RowModelT | None, Exception | None]: Result of row parse.
    """
    cells = {}
    cells.update({"row": row[0].row})
    for key, value in columns_map.items():
        for cell in row:
            if cell.column_letter.lower() == key.lower():
                cells.update({value["name"]: value["parser"](cell)})
    try:
        return (row_model.model_validate(cells), None)
    except ValidationError as exc:
        return (None, exc)

```

Update: Corrected typo in the example.

**Comment by Viicos** [comment]
Try with:

```python
    try:
        return (row_model.model_validate(cells), None)
    except ValidationError as exc:
        return (None, exc)
```

**Comment by samuel-odinware** [comment]
> Try with:
> 
> ```python
>     try:
>         return (row_model.model_validate(cells), None)
>     except ValidationError as exc:
>         return (None, exc)
> ```

This is how I originally had it, but I was trying different approaches and messed that up when I was setting back. Thank you for catching my mistake.

The function still results in a return of `Any` vs `tuple[RowModelT | None, Exception | None]` when it is used. Checking the first element and verifying that it is not `None` also results in `Any` in my editor.

**Comment by Viicos** [comment]
Maybe using `tuple[RowModelT, None] | tuple[None, Exception]` is more suitable.

Also sometimes when having the following:

```python
rv = parse_row(...)  # Your arguments passed to the function
```

Your editor might think `rv` is of type `Any` when you are calling `parse_row` with incorrect types (or even not the right number of arguments). You might want to check if you are calling your function correctly
